content/articles/django-celery-tasks/index.md

```diff
@@ -0,0 +1,339 @@
+   ```html
+   {% extends 'task/base.html' %}
+
+   {% block content %}
+    <h2>Generate Random Users</h2>
+    <form method="post">
+        {% csrf_token %}
+        <table>
+            {{ form }}
+        </table>
+        <button type="submit">Submit</button>
+    </form>
+   {% endblock %}
+   ```
+
+#### Form
+
+In the `task`, create a python file named `form.py` and add the code snippet below:
+
+```python
+from django import forms
+from django.core.validators import MinValueValidator, MaxValueValidator
+
+class GenerateRandomUserForm(forms.Form):
+    total = forms.IntegerField(
+        validators=[
+            MinValueValidator(50),
+            MaxValueValidator(500)
+        ]
+    )
+```
+
+### View
+
+Add the code snippet below into the `views.py` file in the `task`:
+
+```python
+from django.contrib.auth.models import User
+from django.contrib import messages
+from django.views.generic import ListView
+from django.views.generic.edit import FormView
+from django.shortcuts import redirect
+
+from .form import GenerateRandomUserForm
+from .task import create_random_user_accounts
+
+# returns a list of generated user accounts
+class UsersListView(ListView):
+    template_name = 'task/user_list.html'
+    model = User
+
+# A page with the form where we can input the number of accounts to generate
+class GenerateRandomUserView(FormView):
+    template_name = 'task/generate_random_user.html'
+    form_class = GenerateRandomUserForm
+
+    def form_valid(self, form):
+        total = form.cleaned_data.get('total')
+        create_random_user_accounts.delay(total)
+        messages.success(self.request, 'We are generating your random users! Wait a moment and refresh this page.')
+        return redirect('users_list')
+```
+
+In the above code snippet, notice that we did not call the `create_random_user_accounts` method instead we called `create_random_user_accounts.delay(total)`. This instructs Celery to perform the task in a background process.
+
+#### URLs
+
+Update the `urls.py` with the code snippet below.
+
+```python
+from django.conf.urls import url
+from django.contrib import admin
+from django.urls import path
+
+from task.views import GenerateRandomUserView, UsersListView
+
+urlpatterns = [
+    path('admin/', admin.site.urls),
+    url('users/', UsersListView.as_view(), name='users_list'),
+    url('generate/', GenerateRandomUserView.as_view(), name='generate')
+
+]
+```
+
+#### Starting the Celery worker process
+
+Execute the command below to start a Celery worker process.
+ 
+```bash
+celery -A celerytask worker -l info
+```
+
+### Testing
+
+1. Run Django migrations to create the tables in the database by executing the command below.
+
+  ```bash
+  python manage.py migrate
+  ```
+
+2. Start the Django web server by executing the command below.
+
+  ```bash
+  python manage.py runserver
+  ```
+
+3. Start the Celery worker by executing the command below.
+   
+   ```bash
+   celery -A celerytask worker -l info
+   ```
```

Comment:
![image](https://user-images.githubusercontent.com/45763204/126558452-58d236f2-5092-450d-a4ab-b97bb6cde210.png)

I'm getting some errors like this after running this command. any idea why?

Reply:
```bash
[2021-07-24 15:45:12,515: WARNING/MainProcess] No hostname was supplied. Reverting to default 'localhost'
 
 -------------- celery@karen v5.1.2 (sun-harmonics)
--- ***** ----- 
-- ******* ---- Linux-5.8.0-63-generic-x86_64-with-glibc2.29 2021-07-24 15:45:12
- *** --- * --- 
- ** ---------- [config]
- ** ---------- .> app:         celery_task:0x7f4658815b50
- ** ---------- .> transport:   amqp://guest:**@localhost:5672//
- ** ---------- .> results:     disabled://
- *** --- * --- .> concurrency: 4 (prefork)
-- ******* ---- .> task events: OFF (enable -E to monitor tasks in this worker)
--- ***** ----- 
 -------------- [queues]
                .> celery           exchange=celery(direct) key=celery
                

[tasks]
  . task.task.create_random_user_accounts

[2021-07-24 15:45:12,839: INFO/MainProcess] Connected to amqp://guest:**@127.0.0.1:5672//
[2021-07-24 15:45:12,886: INFO/MainProcess] mingle: searching for neighbors
[2021-07-24 15:45:12,931: WARNING/MainProcess] No hostname was supplied. Reverting to default 'localhost'
[2021-07-24 15:45:13,981: INFO/MainProcess] mingle: all alone
[2021-07-24 15:45:14,001: WARNING/MainProcess] /home/karen/.local/lib/python3.8/site-packages/celery/fixups/django.py:203: UserWarning: Using settings.DEBUG leads to a memory
            leak, never use this setting in production environments!
  warnings.warn('''Using settings.DEBUG leads to a memory

[2021-07-24 15:45:14,001: INFO/MainProcess] celery@karen ready.
```

Reply:
@njerikaren I did create a virtual environment both the times I tried. I got the error in spite of creating and using a virtual environment.